### ISSUE TITLE
[JENKINS-50932] `hostPort` only works for one pod at a time

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/kubernetes-plugin#2042]
<p>We are running tests against our environments with a test container and a callback container for the tests in the same pod. The callback container has a hostPort defined to receive the callbacks from the services to be tested. Kubernetes plugin up until version 1.3.3 works with our setup, but starting from 1.4.0 it looks like only one of the three pods running in parallel actually manages to get that hostPort, the other two don't.</p>

<p>The pod definition we are using in the pipeline script looks like this:</p>

<pre>
podTemplate(
  name: podName,
  label: podName,
  instanceCap: 1,
  containers: [
    containerTemplate(
      name: <span class="code-quote">'clitest'</span>,
      image: <span class="code-quote">"<span class="code-keyword">private</span>-docker-repo/platform/clitest:${clitestVersion}"</span>,
      command: <span class="code-quote">'cat'</span>,
      ttyEnabled: <span class="code-keyword">true</span>,
      resourceRequestCpu: <span class="code-quote">'500m'</span>,
      resourceLimitCpu: <span class="code-quote">'2000m'</span>,
      resourceRequestMemory: <span class="code-quote">'1000Mi'</span>,
      resourceLimitMemory: <span class="code-quote">'1000Mi'</span>,
    ),
    containerTemplate(
      name: <span class="code-quote">'rcc'</span>,
      image: <span class="code-quote">'<span class="code-keyword">private</span>-docker-repo/platform/rcc:0.2'</span>,
      command: <span class="code-quote">'/bin/bash'</span>,
      args: <span class="code-quote">'-c <span class="code-quote">"/opt/rcc/start.sh `wget -q -O - http:<span class="code-comment">//169.254.169.254/latest/meta-data/<span class="code-keyword">public</span>-ipv4` '</span> + rccPort + <span class="code-quote">' 30304"</span>'</span>,
</span>      ports: [portMapping(name: <span class="code-quote">'rcc'</span>, containerPort: rccPort, hostPort: rccPort)],
      ttyEnabled: <span class="code-keyword">true</span>,
      resourceRequestCpu: <span class="code-quote">'500m'</span>,
      resourceLimitCpu: <span class="code-quote">'500m'</span>,
      resourceRequestMemory: <span class="code-quote">'2000Mi'</span>,
      resourceLimitMemory: <span class="code-quote">'2000Mi'</span>,
    )
  ],
  volumes: [
    emptyDirVolume(mountPath: <span class="code-quote">'/home/jenkins'</span>, memory: <span class="code-keyword">false</span>),
  ]
)</pre>
<p>For every pod running, we randomly select a port number between a certain range to be used as the hostPort for the `rcc` container and make the process running in the container listen to the same port.</p>

<p>`kubectl describe pod` for three pods started in parallel in a job with is attached. From Jenkins master logs I couldn't find anything meaningful related to this.</p>

<p>Jenkins version: 2.107.2<br/>
Kubernetes plugin version: 1.5.2</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=juhtie01">juhtie01</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50932" target="_blank">`hostPort` only works for one pod at a time</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/csanchez.png" title="csanchez's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=csanchez">csanchez</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: kubernetes-plugin
<li><b>label(s)</b>: regression
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2018-10-15T10:26:59+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>We are running tests against our environments with a test container and a callback container for the tests in the same pod. The callback container has a hostPort defined to receive the callbacks from the services to be tested. Kubernetes plugin up until version 1.3.3 works with our setup, but starting from 1.4.0 it looks like only one of the three pods running in parallel actually manages to get that hostPort, the other two don't.</p>

<p>The pod definition we are using in the pipeline script looks like this:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">podTemplate(
  name: podName,
  label: podName,
  instanceCap: 1,
  containers: [
    containerTemplate(
      name: <span class="code-quote">'clitest'</span>,
      image: <span class="code-quote">"<span class="code-keyword">private</span>-docker-repo/platform/clitest:${clitestVersion}"</span>,
      command: <span class="code-quote">'cat'</span>,
      ttyEnabled: <span class="code-keyword">true</span>,
      resourceRequestCpu: <span class="code-quote">'500m'</span>,
      resourceLimitCpu: <span class="code-quote">'2000m'</span>,
      resourceRequestMemory: <span class="code-quote">'1000Mi'</span>,
      resourceLimitMemory: <span class="code-quote">'1000Mi'</span>,
    ),
    containerTemplate(
      name: <span class="code-quote">'rcc'</span>,
      image: <span class="code-quote">'<span class="code-keyword">private</span>-docker-repo/platform/rcc:0.2'</span>,
      command: <span class="code-quote">'/bin/bash'</span>,
      args: <span class="code-quote">'-c <span class="code-quote">"/opt/rcc/start.sh `wget -q -O - http:<span class="code-comment">//169.254.169.254/latest/meta-data/<span class="code-keyword">public</span>-ipv4` '</span> + rccPort + <span class="code-quote">' 30304"</span>'</span>,
</span>      ports: [portMapping(name: <span class="code-quote">'rcc'</span>, containerPort: rccPort, hostPort: rccPort)],
      ttyEnabled: <span class="code-keyword">true</span>,
      resourceRequestCpu: <span class="code-quote">'500m'</span>,
      resourceLimitCpu: <span class="code-quote">'500m'</span>,
      resourceRequestMemory: <span class="code-quote">'2000Mi'</span>,
      resourceLimitMemory: <span class="code-quote">'2000Mi'</span>,
    )
  ],
  volumes: [
    emptyDirVolume(mountPath: <span class="code-quote">'/home/jenkins'</span>, memory: <span class="code-keyword">false</span>),
  ]
)</pre>
</div></div>
<p>For every pod running, we randomly select a port number between a certain range to be used as the hostPort for the `rcc` container and make the process running in the container listen to the same port.</p>

<p>`kubectl describe pod` for three pods started in parallel in a job with is attached. From Jenkins master logs I couldn't find anything meaningful related to this.</p>

<p>Jenkins version: 2.107.2
Kubernetes plugin version: 1.5.2</p></pre>
</details>
</details>
<details><summary><i>2 attachments</i></summary>

- [clitest-rcc-pods](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/42245/clitest-rcc-pods)
- [clitest-rcc-pods-1.3.3](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/42246/clitest-rcc-pods-1.3.3)
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [PR-340](https://github.com/jenkinsci/kubernetes-plugin/pull/340)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>juhtie01:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50932?focusedId=335970&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-335970">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I attached `clitest-rcc-pods-1.3.3` to show how all the pods have host ports when running with Kubernetes plugin 1.3.3.</p>
</pre>
</details>

</ul>
</details>

<p>I attached `clitest-rcc-pods-1.3.3` to show how all the pods have host ports when running with Kubernetes plugin 1.3.3.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>juhtie01:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50932?focusedId=340135&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-340135">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>PR for a test case to expose this bug: <a href="https://github.com/jenkinsci/kubernetes-plugin/pull/340" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/pull/340</a></p>
</pre>
</details>

</ul>
</details>

<p>PR for a test case to expose this bug: <a href="https://github.com/jenkinsci/kubernetes-plugin/pull/340" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/pull/340</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>juhtie01:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50932?focusedId=341073&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-341073">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Debugged this some more. Here are some logs from the test:</p>

<p> </p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>11.532 [id=61]	FINEST	o.c.j.p.k.p.AbstractInvisibleRunAction2#push: Pushing item hostPort1-z98x6 to action org.csanchez.jenkins.plugins.kubernetes.pipeline.PodTemplateAction@17629438 in run PodsWithHostPorts #1
11.541 [id=61]	FINEST	o.c.j.p.k.p.AbstractInvisibleRunAction2#push: Pushing item kubernetes-plugin-test to action org.csanchez.jenkins.plugins.kubernetes.pipeline.NamespaceAction@5e60add4 in run PodsWithHostPorts #1
hostPort1-z98x6
11.619 [id=61]	FINEST	o.c.j.p.k.p.AbstractInvisibleRunAction2#push: Pushing item hostPort2-pntgz to action org.csanchez.jenkins.plugins.kubernetes.pipeline.PodTemplateAction@17629438 in run PodsWithHostPorts #1
11.624 [id=61]	FINEST	o.c.j.p.k.p.AbstractInvisibleRunAction2#push: Pushing item kubernetes-plugin-test to action org.csanchez.jenkins.plugins.kubernetes.pipeline.NamespaceAction@5e60add4 in run PodsWithHostPorts #1
...
18.915 [id=61]	FINEST	o.c.j.p.k.PodTemplateUtils#combine: Combining pod templates, parent: PodTemplate{, name='hostPort1-z98x6', namespace='kubernetes-plugin-test', label='hostPort1', serviceAccount='jenkins', containers=[ContainerTemplate{name='hyperkube', image='gcr.io/google_containers/hyperkube-amd64:v1.9.7', workingDir='/home/jenkins', command='/bin/cat', ttyEnabled=true, ports=[hostport1,10001]}]}
...
18.916 [id=61]	FINEST	o.c.j.p.k.PodTemplateUtils#combine: Combining pod templates, template: PodTemplate{inheritFrom='hostPort1-z98x6', name='hostPort2-pntgz', namespace='kubernetes-plugin-test', label='hostPort2', serviceAccount='jenkins', containers=[ContainerTemplate{name='hyperkube', image='gcr.io/google_containers/hyperkube-amd64:v1.9.7', workingDir='/home/jenkins', command='/bin/cat', ttyEnabled=true, ports=[hostport2,10002]}]}
18.927 [id=61]	FINEST	o.c.j.p.k.PodTemplateUtils#combine: Pod templates combined: PodTemplate{inheritFrom='hostPort1-z98x6', name='hostPort2-pntgz', namespace='kubernetes-plugin-test', label='hostPort2', serviceAccount='jenkins', containers=[ContainerTemplate{name='hyperkube', image='gcr.io/google_containers/hyperkube-amd64:v1.9.7', workingDir='/home/jenkins', command='/bin/cat', ttyEnabled=true}]}
18.927 [id=61]	FINEST	o.c.j.p.k.PodTemplateUtils#unwrap: Combined parent + template is PodTemplate{inheritFrom='hostPort1-z98x6', name='hostPort2-pntgz', namespace='kubernetes-plugin-test', label='hostPort2', serviceAccount='jenkins', containers=[ContainerTemplate{name='hyperkube', image='gcr.io/google_containers/hyperkube-amd64:v1.9.7', workingDir='/home/jenkins', command='/bin/cat', ttyEnabled=true}]}</pre>
</div></div>
<p> </p>

<p>I looks like the first pod to be pushed to 'org.csanchez.jenkins.plugins.kubernetes.pipeline.PodTemplateAction@17629438' becomes a parent to the second pod, even though the second should not inherit from the first one.</p>

<p>This looks like a duplicate of <a href="https://issues.jenkins-ci.org/browse/JENKINS-50196" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-50196</a></p>
</pre>
</details>

</ul>
</details>

<p>Debugged this some more. Here are some logs from the test:</p>

<p> </p>


```
11.532 [id=61]	FINEST	o.c.j.p.k.p.AbstractInvisibleRunAction2#push: Pushing item hostPort1-z98x6 to action org.csanchez.jenkins.plugins.kubernetes.pipeline.PodTemplateAction@​17629438 in run PodsWithHostPorts #1
11.541 [id=61]	FINEST	o.c.j.p.k.p.AbstractInvisibleRunAction2#push: Pushing item kubernetes-plugin-test to action org.csanchez.jenkins.plugins.kubernetes.pipeline.NamespaceAction@​5e60add4 in run PodsWithHostPorts #1
hostPort1-z98x6
11.619 [id=61]	FINEST	o.c.j.p.k.p.AbstractInvisibleRunAction2#push: Pushing item hostPort2-pntgz to action org.csanchez.jenkins.plugins.kubernetes.pipeline.PodTemplateAction@​17629438 in run PodsWithHostPorts #1
11.624 [id=61]	FINEST	o.c.j.p.k.p.AbstractInvisibleRunAction2#push: Pushing item kubernetes-plugin-test to action org.csanchez.jenkins.plugins.kubernetes.pipeline.NamespaceAction@​5e60add4 in run PodsWithHostPorts #1
...
18.915 [id=61]	FINEST	o.c.j.p.k.PodTemplateUtils#combine: Combining pod templates, parent: PodTemplate{, name='hostPort1-z98x6', namespace='kubernetes-plugin-test', label='hostPort1', serviceAccount='jenkins', containers=[ContainerTemplate{name='hyperkube', image='gcr.io/google_containers/hyperkube-amd64:v1.9.7', workingDir='/home/jenkins', command='/bin/cat', ttyEnabled=true, ports=[hostport1,10001]}]}
...
18.916 [id=61]	FINEST	o.c.j.p.k.PodTemplateUtils#combine: Combining pod templates, template: PodTemplate{inheritFrom='hostPort1-z98x6', name='hostPort2-pntgz', namespace='kubernetes-plugin-test', label='hostPort2', serviceAccount='jenkins', containers=[ContainerTemplate{name='hyperkube', image='gcr.io/google_containers/hyperkube-amd64:v1.9.7', workingDir='/home/jenkins', command='/bin/cat', ttyEnabled=true, ports=[hostport2,10002]}]}
18.927 [id=61]	FINEST	o.c.j.p.k.PodTemplateUtils#combine: Pod templates combined: PodTemplate{inheritFrom='hostPort1-z98x6', name='hostPort2-pntgz', namespace='kubernetes-plugin-test', label='hostPort2', serviceAccount='jenkins', containers=[ContainerTemplate{name='hyperkube', image='gcr.io/google_containers/hyperkube-amd64:v1.9.7', workingDir='/home/jenkins', command='/bin/cat', ttyEnabled=true}]}
18.927 [id=61]	FINEST	o.c.j.p.k.PodTemplateUtils#unwrap: Combined parent + template is PodTemplate{inheritFrom='hostPort1-z98x6', name='hostPort2-pntgz', namespace='kubernetes-plugin-test', label='hostPort2', serviceAccount='jenkins', containers=[ContainerTemplate{name='hyperkube', image='gcr.io/google_containers/hyperkube-amd64:v1.9.7', workingDir='/home/jenkins', command='/bin/cat', ttyEnabled=true}]}
```
<p> </p>

<p>I looks like the first pod to be pushed to 'org.csanchez.jenkins.plugins.kubernetes.pipeline.PodTemplateAction@​17629438' becomes a parent to the second pod, even though the second should not inherit from the first one.</p>

<p>This looks like a duplicate of <a href="https://issues.jenkins-ci.org/browse/JENKINS-50196" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-50196</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>juhtie01:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50932?focusedId=351315&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-351315">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Fixed in 1.12.7.</p>
</pre>
</details>

</ul>
</details>

<p>Fixed in 1.12.7.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `duplicates` from Jira: <a href="https://github.com/jenkinsci/kubernetes-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-50196%22">JENKINS-50196</a>]</i>
